src/schemapack/_internals/normalize.py

```diff
@@ -123,47 +123,131 @@ def denormalize(  # noqa: PLR0912,C901
+        next_embedding_profile = _get_next_embedding_profile(
+            embedding_profile, relation_name
+        )
+        denormalization_context = DenormalizationContext(
+            datapack=datapack,
+            schemapack=schemapack,
+            embedding_profile=next_embedding_profile,
+            resource_blacklist=resource_blacklist,
+        )
+        denormalized_object[relation_name] = _process_recursion(
+            denormalization_context,
+            target_class_name,
+            target_ids,
+        )
 
     return denormalized_object
+
+
+def _initialize_blacklist(
+    embedding_profile, root_class_name, root_resource_id, _resource_blacklist
+) -> dict[ClassName, set[ResourceId]]:
+    """Function to initialize the resource blacklist."""
+    resource_blacklist: dict[ClassName, set[ResourceId]] = defaultdict(set)
+    if embedding_profile is None:
+        resource_blacklist[root_class_name].add(root_resource_id)
+
+    if _resource_blacklist:
+        for class_name, resource_ids in _resource_blacklist.items():
+            resource_blacklist[class_name].update(resource_ids)
+    return resource_blacklist
+
+
+def _get_root_resource(
+    datapack: DataPack, root_class_name: ClassName, root_resource_id: ResourceId
+) -> Resource:
+    """Function to get the root resource from the datapack."""
+    root_class_resources = datapack.resources.get(root_class_name)
+    if not root_class_resources:
+        raise ValidationAssumptionError(context="root class lookup")
+
+    root_resource = root_class_resources.get(root_resource_id)
+    if not root_resource:
+        raise ValidationAssumptionError(context="root resource lookup")
+
+    return root_resource
+
+
+def _get_class_definition(
+    schemapack: SchemaPack, root_class_name: ClassName
+) -> ClassDefinition:
+    """Function to get the class definition from the schemapack."""
+    root_class_definition = schemapack.classes.get(root_class_name)
+    if not root_class_definition:
+        raise RuntimeError(
+            "This is a bug and should not happen. It should be caught by the schemapack"
+            + " spec validation."
+        )
+    return root_class_definition
+
+
+def _should_embed(
+    embedding_profile: Mapping[str, Any] | None, relation_name: str
+) -> bool:
+    """Function to decide whether to embed a relation based on the embedding profile."""
+    if embedding_profile is None:
+        return True
+    value = embedding_profile.get(relation_name)
+    if isinstance(value, bool):
+        return value
+    return isinstance(value, dict)
+
+
+def _get_next_embedding_profile(
+    embedding_profile: Mapping[str, Any] | None, relation_name: str
+) -> dict[str, Any] | None:
+    """Function to get the next embedding profile for a relation."""
+    return (
+        embedding_profile.get(relation_name)
+        if embedding_profile and isinstance(embedding_profile.get(relation_name), dict)
+        else None
+    )
```

Comment:
If the rest is changed, than this could also just return an empty dict for the else case